db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -2528,6 +2511,70 @@ Status DBImpl::AtomicFlushMemTables(
   return s;
 }
 
+Status DBImpl::RetryFlushesForErrorRecovery(FlushReason flush_reason,
+                                            bool wait) {
+  mutex_.AssertHeld();
+  assert(flush_reason == FlushReason::kErrorRecoveryRetryFlush ||
+         flush_reason == FlushReason::kCatchUpAfterErrorRecovery);
+
+  // Collect referenced CFDs. In case of atomic flush, we trust that any
+  // unflushed immutable memtables were cut at a consistent point in time.
+  autovector<ColumnFamilyData*> cfds;
+  for (ColumnFamilyData* cfd : *versions_->GetColumnFamilySet()) {
+    if (!cfd->IsDropped() && cfd->initialized()) {
+      cfd->Ref();
+      cfd->imm()->FlushRequested();
+      cfds.push_back(cfd);
+    }
+  }
+
+  // Submit flush requests for all immutable memtables needing flush.
+  // `flush_memtable_ids` will be populated such that all immutable
+  // memtables eligible for flush are waited on before this function
+  // returns.
+  autovector<uint64_t> flush_memtable_ids;
+  if (immutable_db_options_.atomic_flush) {
+    FlushRequest flush_req;
+    AssignAtomicFlushSeq(cfds);
+    GenerateFlushRequest(cfds, flush_reason, &flush_req);
+    SchedulePendingFlush(flush_req);
+    for (auto& iter : flush_req.cfd_to_max_mem_id_to_persist) {
+      flush_memtable_ids.push_back(iter.second);
+    }
+  } else {
+    for (auto cfd : cfds) {
+      flush_memtable_ids.push_back(cfd->imm()->GetLatestMemTableID());
+      if (cfd->imm()->NumNotFlushed() != 0) {
```

Comment:
Should we do the same check for atomic_flush too? `AtomicFlushMemTables()` calls `SelectColumnFamiliesForAtomicFlush()` which includes this check. If so, we can check this before adding to `cfds` above.